File: cs/src/DataCentric/Platform/Logging/LogEntry.cs
/*
Copyright (C) 2013-present The DataCentric Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

using System;
using System.Collections.Generic;

namespace DataCentric
{
    /// <summary>Log entry consists of formatted message and the original message parameter objects.</summary>
    public class LogEntry
    {
        private string entryText_;

        /// <summary>Truncates message parameters before formatting the message.</summary>
        public LogEntry(LogEntryType entryType, string entrySubType, string message, params object[] messageParams)
        {
            // Truncates message parameters before formatting the message.
            string formattedMessage = LogImpl.FormatMessage(message, messageParams);


            string prefix = entryType.ToString();
            if (!string.IsNullOrEmpty(entrySubType))
            {
                // Append dot delimited subtype if specified
                prefix = string.Concat(prefix, ".", entrySubType);
            }

            // Format entry text
            entryText_ = string.Concat(prefix, ": ", formattedMessage);
        }

        /// <summary>Formatted message with truncated parameter strings.
        /// Use MessageParams to access the original parameter objects.</summary>
        public override string ToString() { return entryText_; }
    }
}

File: py/src/datacentric/platform/logging/log_entry.py
from datacentric.platform.logging.log_entry_type import LogEntryType


class LogEntry:
    """Log entry consists of formatted message and the original message parameter objects."""
    __max_message_param_length = 255

    def __init__(self, entry_type: LogEntryType, entry_sub_type: str, message: str, *message_params: object):
        formatted_message = LogEntry.format_message(message, *message_params)
        prefix = entry_type.name

        if entry_sub_type is not None and entry_sub_type != '':
            prefix = prefix + '.' + entry_sub_type

        self._entry_text = prefix + ': ' + formatted_message

    def __str__(self):
        return self._entry_text

    @staticmethod
    def format_message(message: str, *message_params: object) -> str:
        if len(message_params) > 0:
            formatted_message_params = []
            for arg in message_params:
                arg_str = str(arg)

                # Restrict message length if more than max length
                if len(arg_str) > LogEntry.__max_message_param_length:
                    substring_length: int = (LogEntry.__max_message_param_length - 5) // 2
                    head = arg_str[0:substring_length]
                    tail = arg_str[len(arg_str) - substring_length:len(arg_str)]
                    arg_str = head + ' ... ' + tail
                formatted_message_params.append(arg_str)

            result = message.format(formatted_message_params)

            # If the message ends with four dots (....) because the last argument is truncated
            # while its token is followed by a dot (.), reduce to three dots at the end (...)
            if result.endswith('....'):
                return result[0: len(result) - 1]

            return result

        # Do not perform substitution if no parameters are specified.
        # This will work even if the string contains {} characters
        else:
            return message
